### ISSUE TITLE
No account registered error for IBKR

### ISSUE THREAD

**Issue by benjaminsingleton** [nautechsystems/nautilus_trader#1610]
# Bug Report

It looks like a bug was introduced in 1.191 that is not present in 1.190. I'm not sure yet if this is specific to the Interactive Brokers adapter or whether it is related to other changes.

### Expected Behavior

Able to reconcile existing positions in Interactive Brokers.

### Actual Behavior

```
[1m2024-04-25T02:51:03.955754001Z[0m [1;31m[ERROR] TESTER-001.Portfolio: Cannot update order: no account registered for IB-DU1234567[0m
...
[1m2024-04-25T02:51:03.956474001Z[0m [1;31m[ERROR] TESTER-001.ExecEngine: Cannot handle order fill: no account found for NASDAQ, OrderFilled(instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, venue_order_id=AAPL.NASDAQ, account_id=IB-DU1234567, trade_id=9c8da2fe-726c-454a-84d2-13e76cb8169b, position_id=AAPL.NASDAQ-EXTERNAL, order_side=BUY, order_type=MARKET, last_qty=1200, last_px=185.17 USD, commission=0.00 USD, liquidity_side=TAKER, ts_event=1714013463945493001)[0m
```

Here are some more comprehensive log differences for the same strategy in 1.190 vs 1.191:

**1.190**

```
[1m2024-04-26T01:13:49.187769001Z[0m [INFO] TESTER-001.ExecEngine: Registered ExecutionClient-INTERACTIVE_BROKERS.[0m
[1m2024-04-26T01:13:49.187777001Z[0m [INFO] TESTER-001.ExecEngine: Registered INTERACTIVE_BROKERS for default routing.[0m
```

**1.191**

```
[1m2024-04-25T02:51:03.752673001Z[0m [INFO] TESTER-001.ExecEngine: Registered ExecutionClient-IB[0m
[1m2024-04-25T02:51:03.752682001Z[0m [INFO] TESTER-001.ExecEngine: Registered IB for default routing[0m
```

**1.190**

```
[1m2024-04-26T01:13:49.738580001Z[0m [DEBUG] TESTER-001.ExecEngine: [RECV][RPT] ExecutionMassStatus(client_id=INTERACTIVE_BROKERS, account_id=INTERACTIVE_BROKERS-DU1234567, venue=INTERACTIVE_BROKERS, order_reports={VenueOrderId('AAPL.NASDAQ'): OrderStatusReport(account_id=INTERACTIVE_BROKERS-DU1234567, instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, order_list_id=None, venue_order_id=AAPL.NASDAQ, order_side=BUY, order_type=MARKET, contingency_type=NO_CONTINGENCY, time_in_force=FOK, expire_time=None, order_status=FILLED, price=None, trigger_price=None, trigger_type=NO_TRIGGER, limit_offset=None, trailing_offset=None, trailing_offset_type=NO_TRAILING_OFFSET, quantity=1_200, filled_qty=1_200, leaves_qty=0, display_qty=None, avg_px=185.17, post_only=False, reduce_only=False, cancel_reason=None, report_id=5d192a1f-666f-4373-a995-8f2e046546a5, ts_accepted=1714094029440501001, ts_triggered=0, ts_last=1714094029440501001, ts_init=1714094029440501001), VenueOrderId('BTC/USD.PAXOS'): OrderStatusReport(account_id=INTERACTIVE_BROKERS-DU1234567, instrument_id=BTC/USD.PAXOS, client_order_id=BTC/USD.PAXOS, order_list_id=None, venue_order_id=BTC/USD.PAXOS, order_side=BUY, order_type=MARKET, contingency_type=NO_CONTINGENCY, time_in_force=FOK, expire_time=None, order_status=FILLED, price=None, trigger_price=None, trigger_type=NO_TRIGGER, limit_offset=None, trailing_offset=None, trailing_offset_type=NO_TRAILING_OFFSET, quantity=0.05159349, filled_qty=0.05159349, leaves_qty=0.00000000, display_qty=None, avg_px=42908.56, post_only=False, reduce_only=False, cancel_reason=None, report_id=ac1768c3-eabd-4fbc-a65f-c39114ccccf0, ts_accepted=1714094029440501001, ts_triggered=0, ts_last=1714094029440501001, ts_init=1714094029440501001)}, fill_reports={}, position_reports={InstrumentId('AAPL.NASDAQ'): [PositionStatusReport(account_id=INTERACTIVE_BROKERS-DU1234567, instrument_id=AAPL.NASDAQ, venue_position_id=None, position_side=LONG, quantity=1_200, signed_decimal_qty=1200, report_id=7a4663d4-c8c2-4f9a-85db-98d5daf9cc51, ts_last=1714094029682576001, ts_init=1714094029682576002)], InstrumentId('BTC/USD.PAXOS'): [PositionStatusReport(account_id=INTERACTIVE_BROKERS-DU1234567, instrument_id=BTC/USD.PAXOS, venue_position_id=None, position_side=LONG, quantity=0.05159349, signed_decimal_qty=0.05159349, report_id=d9279b93-9c27-40ae-98af-0e0c9d581f9f, ts_last=1714094029737241001, ts_init=1714094029737241002)]}, report_id=0c4dd589-5e0f-4054-8723-fcf315cc18a4, ts_init=1714094029435012001).[0m
[1m2024-04-26T01:13:49.738598001Z[0m [94m[INFO] TESTER-001.ExecEngine: Reconciling ExecutionMassStatus for INTERACTIVE_BROKERS.[0m
[1m2024-04-26T01:13:49.738607001Z[0m [94m[INFO] TESTER-001.ExecEngine: Reconciling order for ClientOrderId('AAPL.NASDAQ')...[0m
[1m2024-04-26T01:13:49.738655001Z[0m [94m[INFO] TESTER-001.ExecEngine: Generating order ClientOrderId('AAPL.NASDAQ')[0m
[1m2024-04-26T01:13:49.738903001Z[0m [DEBUG] TESTER-001.ExecEngine: Generated OrderInitialized(instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, side=BUY, type=MARKET, quantity=1_200, time_in_force=GTC, post_only=False, reduce_only=False, quote_quantity=False, options={'expire_time_ns': 0}, emulation_trigger=NO_TRIGGER, trigger_instrument_id=None, contingency_type=NO_CONTINGENCY, order_list_id=None, linked_order_ids=None, parent_order_id=None, exec_algorithm_id=None, exec_algorithm_params=None, exec_spawn_id=None, tags=EXTERNAL).[0m
[1m2024-04-26T01:13:49.738926001Z[0m [DEBUG] TESTER-001.Cache: Added MarketOrder(BUY 1_200 AAPL.NASDAQ MARKET GTC, status=INITIALIZED, client_order_id=AAPL.NASDAQ, venue_order_id=None, position_id=None, tags=EXTERNAL).[0m
[1m2024-04-26T01:13:49.738996001Z[0m [DEBUG] TESTER-001.ExecEngine: Generated OrderAccepted(instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, venue_order_id=AAPL.NASDAQ, account_id=INTERACTIVE_BROKERS-DU1234567, ts_event=1714094029440501001).[0m
[1m2024-04-26T01:13:49.739701001Z[0m [1;33m[WARN] TESTER-001.ExecEngine: Generated inferred OrderFilled(instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, venue_order_id=AAPL.NASDAQ, account_id=INTERACTIVE_BROKERS-DU1234567, trade_id=3646bbec-22a1-478d-ba98-5285f1e6938c, position_id=AAPL.NASDAQ-EXTERNAL, order_side=BUY, order_type=MARKET, last_qty=1200, last_px=185.17 USD, commission=0.00 USD, liquidity_side=TAKER, ts_event=1714094029440501001).[0m
[1m2024-04-26T01:13:49.739857001Z[0m [DEBUG] TESTER-001.Cache: Indexed PositionId('AAPL.NASDAQ-EXTERNAL'), client_order_id=AAPL.NASDAQ, strategy_id=EXTERNAL).[0m
[1m2024-04-26T01:13:49.739866001Z[0m [DEBUG] TESTER-001.Cache: Added Position(id=AAPL.NASDAQ-EXTERNAL, strategy_id=EXTERNAL).[0m
[1m2024-04-26T01:13:49.740057001Z[0m [INFO] TESTER-001.Portfolio: AAPL.NASDAQ net_position=1200[0m
[1m2024-04-26T01:13:49.740083001Z[0m [DEBUG] TESTER-001.Portfolio: Cannot calculate unrealized PnL: no prices for AAPL.NASDAQ.[0m

```

**1.191**
```
[1m2024-04-25T02:51:03.953953001Z[0m [DEBUG] TESTER-001.ExecEngine: [RECV][RPT] ExecutionMassStatus(client_id=IB, account_id=IB-DU1234567, venue=INTERACTIVE_BROKERS, order_reports={VenueOrderId('AAPL.NASDAQ'): OrderStatusReport(account_id=IB-DU1234567, instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, order_list_id=None, venue_order_id=AAPL.NASDAQ, order_side=BUY, order_type=MARKET, contingency_type=NO_CONTINGENCY, time_in_force=FOK, expire_time=None, order_status=FILLED, price=None, trigger_price=None, trigger_type=NO_TRIGGER, limit_offset=None, trailing_offset=None, trailing_offset_type=NO_TRAILING_OFFSET, quantity=1_200, filled_qty=1_200, leaves_qty=0, display_qty=None, avg_px=185.17, post_only=False, reduce_only=False, cancel_reason=None, report_id=61624aac-be41-4b58-91cd-26f57aba6dd0, ts_accepted=1714013463945493001, ts_triggered=0, ts_last=1714013463945493001, ts_init=1714013463945493001), VenueOrderId('BTC/USD.PAXOS'): OrderStatusReport(account_id=IB-DU1234567, instrument_id=BTC/USD.PAXOS, client_order_id=BTC/USD.PAXOS, order_list_id=None, venue_order_id=BTC/USD.PAXOS, order_side=BUY, order_type=MARKET, contingency_type=NO_CONTINGENCY, time_in_force=FOK, expire_time=None, order_status=FILLED, price=None, trigger_price=None, trigger_type=NO_TRIGGER, limit_offset=None, trailing_offset=None, trailing_offset_type=NO_TRAILING_OFFSET, quantity=0.05159349, filled_qty=0.05159349, leaves_qty=0.00000000, display_qty=None, avg_px=42908.56, post_only=False, reduce_only=False, cancel_reason=None, report_id=7a5fe544-6523-4fdf-8c87-3b267eed1106, ts_accepted=1714013463945493001, ts_triggered=0, ts_last=1714013463945493001, ts_init=1714013463945493001)}, fill_reports={}, position_reports={InstrumentId('AAPL.NASDAQ'): [PositionStatusReport(account_id=IB-DU1234567, instrument_id=AAPL.NASDAQ, venue_position_id=None, position_side=LONG, quantity=1_200, signed_decimal_qty=1200, report_id=2c9c0f0f-51f4-4419-b15d-0063a53e7506, ts_last=1714013463950672001, ts_init=1714013463950672002)], InstrumentId('BTC/USD.PAXOS'): [PositionStatusReport(account_id=IB-DU1234567, instrument_id=BTC/USD.PAXOS, venue_position_id=None, position_side=LONG, quantity=0.05159349, signed_decimal_qty=0.05159349, report_id=f4abdcf7-7e6a-4ea2-85db-1386a36089d2, ts_last=1714013463953082001, ts_init=1714013463953082002)]}, report_id=975179f4-50ad-47ca-a3c9-640414042269, ts_init=1714013463943010001)[0m
[1m2024-04-25T02:51:03.953964001Z[0m [94m[INFO] TESTER-001.ExecEngine: Reconciling ExecutionMassStatus for INTERACTIVE_BROKERS[0m
[1m2024-04-25T02:51:03.953967001Z[0m [94m[INFO] TESTER-001.ExecEngine: Reconciling order for ClientOrderId('AAPL.NASDAQ')[0m
[1m2024-04-25T02:51:03.953970001Z[0m [94m[INFO] TESTER-001.ExecEngine: Generating order ClientOrderId('AAPL.NASDAQ')[0m
[1m2024-04-25T02:51:03.954565001Z[0m [DEBUG] TESTER-001.ExecEngine: Generated OrderInitialized(instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, side=BUY, type=MARKET, quantity=1_200, time_in_force=GTC, post_only=False, reduce_only=False, quote_quantity=False, options={'expire_time_ns': 0}, emulation_trigger=NO_TRIGGER, trigger_instrument_id=None, contingency_type=NO_CONTINGENCY, order_list_id=None, linked_order_ids=None, parent_order_id=None, exec_algorithm_id=None, exec_algorithm_params=None, exec_spawn_id=None, tags=EXTERNAL)[0m
[1m2024-04-25T02:51:03.954595001Z[0m [DEBUG] TESTER-001.Cache: Added MarketOrder(BUY 1_200 AAPL.NASDAQ MARKET GTC, status=INITIALIZED, client_order_id=AAPL.NASDAQ, venue_order_id=None, position_id=None, tags=EXTERNAL)[0m
[1m2024-04-25T02:51:03.954919001Z[0m [DEBUG] TESTER-001.ExecEngine: Generated OrderAccepted(instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, venue_order_id=AAPL.NASDAQ, account_id=IB-DU1234567, ts_event=1714013463945493001)[0m
[1m2024-04-25T02:51:03.955754001Z[0m [1;31m[ERROR] TESTER-001.Portfolio: Cannot update order: no account registered for IB-DU1234567[0m
[1m2024-04-25T02:51:03.956433001Z[0m [1;33m[WARN] TESTER-001.ExecEngine: Generated inferred OrderFilled(instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, venue_order_id=AAPL.NASDAQ, account_id=IB-DU1234567, trade_id=9c8da2fe-726c-454a-84d2-13e76cb8169b, position_id=AAPL.NASDAQ-EXTERNAL, order_side=BUY, order_type=MARKET, last_qty=1200, last_px=185.17 USD, commission=0.00 USD, liquidity_side=TAKER, ts_event=1714013463945493001)[0m
[1m2024-04-25T02:51:03.956454001Z[0m [1;31m[ERROR] TESTER-001.Portfolio: Cannot update order: no account registered for IB-DU1234567[0m
[1m2024-04-25T02:51:03.956474001Z[0m [1;31m[ERROR] TESTER-001.ExecEngine: Cannot handle order fill: no account found for NASDAQ, OrderFilled(instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, venue_order_id=AAPL.NASDAQ, account_id=IB-DU1234567, trade_id=9c8da2fe-726c-454a-84d2-13e76cb8169b, position_id=AAPL.NASDAQ-EXTERNAL, order_side=BUY, order_type=MARKET, last_qty=1200, last_px=185.17 USD, commission=0.00 USD, liquidity_side=TAKER, ts_event=1714013463945493001)[0m

```

### Steps to Reproduce the Problem

1. Add positions in IBKR
2. Start NT

### Specifications

- OS platform: MacOS Sonoma 14.3.1
- Python version: 3.10
- `nautilus_trader` version: 1.191

**Comment by rsmb7z** [comment]
@benjaminsingleton 
Because factory name is now overriding hence special treatment of Portfolio might not be working. Do you still experience if you make any of the following changes?
a) Set `node.portfolio.set_specific_venue(Venue("IB"))` (because factory name according to your logs is 'IB')
b) Set factory name to `INTERACTIVE_BROKERS` which is the value of IB_VENUE

**Comment by benjaminsingleton** [comment]
This got my code working again: 

> b) Set factory name to INTERACTIVE_BROKERS which is the value of IB_VENUE

But this didn't:

> a) Set node.portfolio.set_specific_venue(Venue("IB")) (because factory name according to your logs is 'IB')


```
2024-04-26T02:19:25.220031001Z [INFO] TESTER-001.ExecEngine: Reconciling ExecutionMassStatus for INTERACTIVE_BROKERS
2024-04-26T02:19:25.220034001Z [INFO] TESTER-001.ExecEngine: Reconciling order for ClientOrderId('AAPL.NASDAQ')
2024-04-26T02:19:25.220035001Z [INFO] TESTER-001.ExecEngine: Generating order ClientOrderId('AAPL.NASDAQ')
2024-04-26T02:19:25.220506001Z [DEBUG] TESTER-001.ExecEngine: Generated OrderInitialized(instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, side=BUY, type=MARKET, quantity=1_200, time_in_force=GTC, post_only=False, reduce_only=False, quote_quantity=False, options={'expire_time_ns': 0}, emulation_trigger=NO_TRIGGER, trigger_instrument_id=None, contingency_type=NO_CONTINGENCY, order_list_id=None, linked_order_ids=None, parent_order_id=None, exec_algorithm_id=None, exec_algorithm_params=None, exec_spawn_id=None, tags=EXTERNAL)
2024-04-26T02:19:25.220644001Z [DEBUG] TESTER-001.Cache: Added MarketOrder(BUY 1_200 AAPL.NASDAQ MARKET GTC, status=INITIALIZED, client_order_id=AAPL.NASDAQ, venue_order_id=None, position_id=None, tags=EXTERNAL)
2024-04-26T02:19:25.220939001Z [DEBUG] TESTER-001.ExecEngine: Generated OrderAccepted(instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, venue_order_id=AAPL.NASDAQ, account_id=IB-DU1234567, ts_event=1714097965209516001)
2024-04-26T02:19:25.221667001Z [ERROR] TESTER-001.Portfolio: Cannot update order: no account registered for IB-DU1234567
2024-04-26T02:19:25.222006001Z [WARN] TESTER-001.ExecEngine: Generated inferred OrderFilled(instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, venue_order_id=AAPL.NASDAQ, account_id=IB-DU1234567, trade_id=0edb8a82-65f9-4db6-b900-9428c46b5fb3, position_id=AAPL.NASDAQ-EXTERNAL, order_side=BUY, order_type=MARKET, last_qty=1200, last_px=185.17 USD, commission=0.00 USD, liquidity_side=TAKER, ts_event=1714097965209516001)
2024-04-26T02:19:25.222023001Z [ERROR] TESTER-001.Portfolio: Cannot update order: no account registered for IB-DU1234567
2024-04-26T02:19:25.222044001Z [ERROR] TESTER-001.ExecEngine: Cannot handle order fill: no account found for NASDAQ, OrderFilled(instrument_id=AAPL.NASDAQ, client_order_id=AAPL.NASDAQ, venue_order_id=AAPL.NASDAQ, account_id=IB-DU1234567, trade_id=0edb8a82-65f9-4db6-b900-9428c46b5fb3, position_id=AAPL.NASDAQ-EXTERNAL, order_side=BUY, order_type=MARKET, last_qty=1200, last_px=185.17 USD, commission=0.00 USD, liquidity_side=TAKER, ts_event=1714097965209516001)
2024-04-26T02:19:25.222053001Z [INFO] TESTER-001.ExecEngine: Reconciling order for ClientOrderId('BTC/USD.PAXOS')
2024-04-26T02:19:25.222057001Z [INFO] TESTER-001.ExecEngine: Generating order ClientOrderId('BTC/USD.PAXOS')
2024-04-26T02:19:25.222086001Z [DEBUG] TESTER-001.ExecEngine: Generated OrderInitialized(instrument_id=BTC/USD.PAXOS, client_order_id=BTC/USD.PAXOS, side=BUY, type=MARKET, quantity=0.05159349, time_in_force=GTC, post_only=False, reduce_only=False, quote_quantity=False, options={'expire_time_ns': 0}, emulation_trigger=NO_TRIGGER, trigger_instrument_id=None, contingency_type=NO_CONTINGENCY, order_list_id=None, linked_order_ids=None, parent_order_id=None, exec_algorithm_id=None, exec_algorithm_params=None, exec_spawn_id=None, tags=EXTERNAL)
2024-04-26T02:19:25.222090001Z [DEBUG] TESTER-001.Cache: Added MarketOrder(BUY 0.05159349 BTC/USD.PAXOS MARKET GTC, status=INITIALIZED, client_order_id=BTC/USD.PAXOS, venue_order_id=None, position_id=None, tags=EXTERNAL)
2024-04-26T02:19:25.222104001Z [DEBUG] TESTER-001.ExecEngine: Generated OrderAccepted(instrument_id=BTC/USD.PAXOS, client_order_id=BTC/USD.PAXOS, venue_order_id=BTC/USD.PAXOS, account_id=IB-DU1234567, ts_event=1714097965209516001)
2024-04-26T02:19:25.222107001Z [ERROR] TESTER-001.Portfolio: Cannot update order: no account registered for IB-DU1234567
2024-04-26T02:19:25.222133001Z [WARN] TESTER-001.ExecEngine: Generated inferred OrderFilled(instrument_id=BTC/USD.PAXOS, client_order_id=BTC/USD.PAXOS, venue_order_id=BTC/USD.PAXOS, account_id=IB-DU1234567, trade_id=419e00e0-5dfa-4559-9724-6e35e22d325f, position_id=BTC/USD.PAXOS-EXTERNAL, order_side=BUY, order_type=MARKET, last_qty=0.05159349, last_px=42908.56 USD, commission=0.00 USD, liquidity_side=TAKER, ts_event=1714097965209516001)
2024-04-26T02:19:25.222140001Z [ERROR] TESTER-001.Portfolio: Cannot update order: no account registered for IB-DU1234567
```

**Comment by rsmb7z** [comment]
@cjdsellers 
It is because the name doesn't match [here](https://github.com/nautechsystems/nautilus_trader/blob/2423a3db9c57c3f44a2786d2d96490ec3e611567/nautilus_trader/live/node_builder.py#L267) after 'Implement custom client IDs'. Because the name can be customized now and dynamically setting may not work in other cases as well (at the moment) i.e if user tries run more than one-adapters, it may be less confusing if we require the user to explicity `set_specific_venue` as specified in example.

**Comment by rsmb7z** [comment]
> But then following onto [L268](https://github.com/nautechsystems/nautilus_trader/blob/2423a3db9c57c3f44a2786d2d96490ec3e611567/nautilus_trader/live/node_builder.py#L268) we set a specific venue of `Venue("InteractiveBrokers")`. Will this still work with multiple Interactive Brokers clients?

The problem I believe is we only have one Portfolio, while it may not raise any errors, multiple ExecCients would be sending their own balance updates and calculations maynot be as expected within Portfolio. Some more info is tracked in #1051.





> we parse the custom `name` [here](https://github.com/nautechsystems/nautilus_trader/blob/2423a3db9c57c3f44a2786d2d96490ec3e611567/nautilus_trader/live/node_builder.py#L223).
> 
> Then I think we shouldn't rely on this hard coded string (I note the "temporary" comment there too), but should check if `factory` is an instance of either `InteractiveBrokersDataExecClientFactory` or `InteractiveBrokersLiveExecClientFactory`?

Both of the L202 and L268 would be removed and for ExecClient dynamic addition can be done when the Factory is IB, if you prefer that way.

**Comment by benjaminsingleton** [comment]
In `node_builder.py`, should we check if `name == 'INTERACTIVE_BROKERS'` instead of `InteractiveBrokers` so it matches `IB_VENUE` [here](https://github.com/nautechsystems/nautilus_trader/blob/develop/nautilus_trader/adapters/interactive_brokers/common.py#L26)?

**Comment by rsmb7z** [comment]
I have pushed the fix in #1616.

**Comment by benjaminsingleton** [comment]
Thank you!
